x-pack/test/api_integration/deployment_agnostic/services/role_scoped_supertest.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  RoleCredentials,
+  SupertestWithoutAuthProviderType,
+  SamlAuthProviderType,
+} from '@kbn/ftr-common-functional-services';
+import { Test } from 'supertest';
+import { DeploymentAgnosticFtrProviderContext } from '../ftr_provider_context';
+
+export interface RequestHeadersOptions {
+  withInternalHeaders?: boolean;
+  withCommonHeaders?: boolean;
+  withCustomHeaders?: Record<string, string>;
+}
+
+export class SupertestWithRoleScope {
+  private roleAuthc: RoleCredentials | null;
+  private supertestWithoutAuth: SupertestWithoutAuthProviderType;
+  private samlAuth: SamlAuthProviderType;
+  private options: RequestHeadersOptions;
+  private destroyed = false;
+
+  constructor(
+    roleAuthc: RoleCredentials,
+    supertestWithoutAuth: SupertestWithoutAuthProviderType,
+    samlAuth: SamlAuthProviderType,
+    options: RequestHeadersOptions
+  ) {
+    this.roleAuthc = roleAuthc;
+    this.supertestWithoutAuth = supertestWithoutAuth;
+    this.samlAuth = samlAuth;
+    this.options = options;
+  }
+
+  async destroy() {
+    if (this.roleAuthc) {
+      await this.samlAuth.invalidateM2mApiKeyWithRoleScope(this.roleAuthc);
+      this.roleAuthc = null;
+      this.destroyed = true;
+    }
+  }
+
+  private addHeaders(agent: Test): Test {
+    const { withInternalHeaders, withCommonHeaders, withCustomHeaders } = this.options;
+
+    if (!this.roleAuthc) {
+      throw new Error(
+        `'roleAuthc' is not set, most likely it was invalidated or the instance was destroyed.`
+      );
+    }
```

Comment:
__nit:__ it seems we can say quite certainly that the instance was destroyed since there is no other way to nullify `roleAuthc`, right? 🙂

```suggestion
    if (!this.roleAuthc) {
      throw new Error('The instance has already been destroyed.');
    }
```

Or something along these lines.